platform/source/mbed_retarget.cpp

```diff
@@ -327,27 +223,29 @@ MBED_WEAK FileHandle *mbed::mbed_override_console(int fd)
     return NULL;
 }
 
+#if MBED_CONF_TARGET_CONSOLE_UART && DEVICE_SERIAL
+static void set_console_flow_control(SerialBase *console)
+{
+#if   CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTS
+    console->set_flow_control(SerialBase::RTS, STDIO_UART_RTS, NC);
+#elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_CTS
+    console->set_flow_control(SerialBase::CTS, NC, STDIO_UART_CTS);
+#elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTSCTS
+    console->set_flow_control(SerialBase::RTSCTS, STDIO_UART_RTS, STDIO_UART_CTS);
+#endif
+}
+#endif // MBED_CONF_TARGET_CONSOLE_UART && DEVICE_SERIAL
+
 static FileHandle *default_console()
 {
 #if MBED_CONF_TARGET_CONSOLE_UART && DEVICE_SERIAL
 
 #  if MBED_CONF_PLATFORM_STDIO_BUFFERED_SERIAL
-    static const serial_pinmap_t console_pinmap = get_uart_pinmap(STDIO_UART_TX, STDIO_UART_RX);
-    static UARTSerial console(console_pinmap, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
-#   if   CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTS
-    static const serial_fc_pinmap_t fc_pinmap = get_uart_fc_pinmap(STDIO_UART_RTS, NC);
-    console.serial_set_flow_control(SerialBase::RTS, fc_pinmap);
-#   elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_CTS
-    static const serial_fc_pinmap_t fc_pinmap = get_uart_fc_pinmap(NC, STDIO_UART_CTS);
-    console.serial_set_flow_control(SerialBase::CTS, fc_pinmap);
-#   elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTSCTS
-    static const serial_fc_pinmap_t fc_pinmap = get_uart_fc_pinmap(STDIO_UART_RTS, STDIO_UART_CTS);
-    console.serial_set_flow_control(SerialBase::RTSCTS, fc_pinmap);
-#   endif
+    static UARTSerial console(STDIO_UART_TX, STDIO_UART_RX, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
 #  else
-    static const serial_pinmap_t console_pinmap = get_uart_pinmap(STDIO_UART_TX, STDIO_UART_RX);
-    static DirectSerial console(console_pinmap, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
+    static UnbufferedSerial console(STDIO_UART_TX, STDIO_UART_RX, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
 #  endif
+    set_console_flow_control(&console);
```

Comment:
This won't work with buffered serial - `UARTSerial` doesn't expose its `SerialBase`. That's mainly to prevent people messing with the data flow, but also stops you reaching the flow control calls. 

You could deal with it neatly enough by using a typedef - typedef `ConsoleSerial` or something to either `UnbufferedSerial` or `UARTSerial`, and then `set_console_flow_control` could take that. Or you could make `set_console_flow_control` be a template.

Reply:
[Done](https://github.com/ARMmbed/mbed-os/compare/01d72efcb0b950892e446c238d3bcace8528b6a9..01dac5b91aa1541aaf03a45701d43130fe3d5d90).